### ISSUE TITLE
Feature request: early stopping for model training

### ISSUE THREAD

**Issue by gmittal** [google/flax#728]
Opening this feature request based on #711.

### Description of the feature to be implemented
Object to stop training when a monitored metric (e.g. validation loss, etc.) has stopped improving.

### Specific points to consider
- Whether this should live in `flax.training`
- Should this stop the training loop early or should it stop checkpointing early (but continue the train loop)? If so, how should it do that?

### Reference implementations in other frameworks
- https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/python/keras/callbacks.py#L1559-L1690
- https://pytorch-lightning.readthedocs.io/en/latest/_modules/pytorch_lightning/callbacks/early_stopping.html#EarlyStopping

**Comment by avital** [comment]
Curious how this relates to [CLU](https://github.com/google/CommonLoopUtils) library (cc @andsteing @Marvin182)

**Comment by marcvanzee** [comment]
Copying @jheek comments from PR #711 for reference:

"My suggestion would be to consider the following though:
We generally try to avoid "inversion of control" APIs (a somewhat vague term). In this case the EarlyStopping class takes control of checkpointing even though it only adds the logic to decide whether we should stop or not.
I think a proposal that decouples checkpointing from the early stopping criteria (I imagine there are many criteria like this) is more likely to be accepted."

**Comment by avital** [comment]
My main question is whether we need `EarlyStopping` to itself call `save_checkpoint`. Is there a simple way to devise an API that simply returns true/false as to whether training should stop? Then adding it into an existing training loop is just adding two lines regardless of how that training loop was handling checkpointing.

What do you think @gmittal? Would you be OK trying this out and showing a sample diff of how you'd also add this to one of our existing training loops (doesn't really matter which)?

**Comment by marcvanzee** [comment]
It would be useful to shows how your code works somewhere, perhaps in an example, or just with a test.

**Comment by gmittal** [comment]
@avital: One thing that might be worth exploring is having `EarlyStopping` as an iterator that we wrap around the training loop and have that terminate early when the validation metric does not improve:

```python
epoch_iterator = EarlyStopping(range(epochs))
for step in epoch_iterator:
    optimizer = train_step(...)
    validation_metric = ...

    # It would be nice to not save checkpoints that do not improve
    ckpt = flax.training.save_checkpoint(...)

    epoch_iterator.update(validation_metric)
```

I'll put together a couple of examples and post them here soon.

**Comment by gmittal** [comment]
@avital @marcvanzee I've rewritten `EarlyStopping` as an iterator and have pushed some simple tests as well: https://github.com/google/flax/pull/711/files

**Comment by jheek** [comment]
What is the benefit of writing this as an iterator?

I would expect a simple api that can do 3 things:
1) `did_improve = early_stop.update(...)` - incorporate eval metric and return True if it's the best one so far.
2) `early_stop.should_stop` -- tells me if the early stop criteria is met.
3) EarlyStop itself is a flax.struct.dataclass which can be part of a checkpoint.

The train loop would look something like this:
```
early_stop = EarlyStop(...)
for step in range(epochs):
    optimizer = train_step(...)
    validation_metric = ...

    did_improve = early_stop.update(validation_metric)
    if did_improve:
      state = (optimizer, early_stop, ...)
      save_checkpoint(state, prefix='best_')

    if early_stop.should_stop:
      logging.info('early stop....')
      # cleanup
      break
```

**Comment by gmittal** [comment]
Thanks for the feedback. I've updated the object definition and tests accordingly.
